spring-cloud-stream-binder-kafka/src/main/java/org/springframework/cloud/stream/binder/kafka/KafkaBinderMetrics.java

```diff
@@ -114,33 +117,51 @@ public void setTimeout(int timeout) {
 	}
 
 	@Override
-	public void bindTo(MeterRegistry registry) {
-
-		this.scheduler = Executors.newScheduledThreadPool(this.binder.getTopicsInUse().size());
-
-		for (Map.Entry<String, KafkaMessageChannelBinder.TopicInformation> topicInfo : this.binder
-				.getTopicsInUse().entrySet()) {
-
-			if (!topicInfo.getValue().isConsumerTopic()) {
-				continue;
-			}
-
-			String topic = topicInfo.getKey();
-			String group = topicInfo.getValue().getConsumerGroup();
-
-			final Gauge register = Gauge.builder(OFFSET_LAG_METRIC_NAME, this,
-					(o) -> computeAndGetUnconsumedMessages(topic, group)).tag("group", group)
-					.tag("topic", topic)
-					.description("Unconsumed messages for a particular group and topic")
-					.register(registry);
-
-			if (!(register instanceof NoopGauge)) {
-				//Schedule a task to compute the unconsumed messages for this group/topic every minute.
-				this.scheduler.scheduleWithFixedDelay(computeUnconsumedMessagesRunnable(topic, group, this.metadataConsumers),
-						10, DELAY_BETWEEN_TASK_EXECUTION, TimeUnit.SECONDS);
-			}
-		}
-	}
+    public void bindTo(MeterRegistry registry) {
+        /**
+         * We can't just replace one scheduler with another.
+         * Before and even after the old one is gathered by GC, it's threads still exist, consume memory and CPU resources to switch contexts.
+         * Theoretically, as a result of processing n topics, there will be about (1+n)*n/2 threads simultaneously at the same time.
+         */
+        if (this.scheduler != null) {
+            LOG.info("Try to shutdown the old scheduler with " + ((ScheduledThreadPoolExecutor) scheduler).getPoolSize() + " threads");
+            this.scheduler.shutdown();
```

Comment:
Why not just `shutdownNow()`? The scheduler thread could be waiting in `consumer.endOffsets()`; not sure 1 second is long enough if you don't want to interrupt.

Reply:
To be honest, I prefer not to interrupt code that is not completely under my control and that doesn't handle this signal in a very clear for me way. So I suggest just removing the wait. In my OOM case it is enough, that is, at least we will make it "not worse".